src/hotspot/share/opto/mulnode.cpp

```diff
@@ -2052,94 +2052,88 @@ const Type* RotateRightNode::Value(PhaseGVN* phase) const {
   }
 }
 
-// Given an expression (AndX shift mask) or (AndX mask shift),
+// Returns a lower bound of the number of trailing zeros in expr.
+jint MulNode::AndIL_min_trailing_zeros(PhaseGVN* phase, Node* expr, BasicType bt) {
+  expr = expr->uncast();
+  if (expr == nullptr) {
+    return 0;
+  }
+  const TypeInteger* type = phase->type(expr)->isa_integer(bt);
+  if (type == nullptr) {
+    return 0;
+  }
+
+  if (type->is_con()) {
+    long con = type->get_con_as_long(type->basic_type());
+    return con == 0L ? 0 : count_trailing_zeros(con);
+  }
+
+  if (expr->Opcode() == Op_ConvI2L) {
+    expr = expr->in(1);
+    if (expr == nullptr) {
+      return 0;
+    }
+    expr = expr->uncast();
+    if (expr == nullptr) {
+      return 0;
+    }
+    type = phase->type(expr)->isa_int();
+  }
+
+  if (expr->Opcode() == Op_LShift(type->basic_type())) {
+    Node* rhs = expr->in(2);
+    if (rhs == nullptr) {
+      return 0;
+    }
+    const TypeInt* rhs_t = phase->type(rhs)->isa_int();
+    if (!rhs_t || !rhs_t->is_con()) {
+      return 0;
+    }
+    return rhs_t->get_con() & ((type->isa_int() ? BitsPerJavaInteger : BitsPerJavaLong) - 1);
```

Comment:
If you reassign `bt`, you can do `type2aelementbytes(bt)`, which IMO is clearer.

Reply:
Done.